connman/vpn/plugins/l2tp.c

```diff
@@ -125,12 +125,40 @@ struct {
 static DBusConnection *connection;
 
 struct l2tp_private_data {
+	struct vpn_provider *provider;
 	struct connman_task *task;
 	char *if_name;
 	vpn_provider_connect_cb_t cb;
 	void *user_data;
 };
 
+static void l2tp_connect_done(struct l2tp_private_data *data, int err)
+{
+	vpn_provider_connect_cb_t cb;
+	void *user_data;
+
+	if (!data || !data->cb)
+		return;
+
+	/* Ensure that callback is called only once */
+	cb = data->cb;
+	user_data = data->user_data;
+	data->cb = NULL;
+	data->user_data = NULL;
+	cb(data->provider, user_data, err);
+}
+
+static void free_private_data(struct l2tp_private_data *data)
```

Comment:
I wish also allocations were encapsulated in functions... would not need to go back and forth between here and inlined allocation code to see that what is allocated also gets released. But I guess this is how it is, matter of opinion,  and does not affect correctness of the code.

Reply:
There is a task where a library/utils for the VPNs is to be created and this is one part of it as these private data structs across every plugin have a lot of similarities and mostly same content. These need a major refactoring then, oh well, upstream approved these l2tp and pptp changes as is.